File: c++/123. Best Time to Buy and Sell Stock III.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        auto &a = prices;
        int n = a.size();
        if (n < 2) return 0;
        
        vector<int> v1(n, 0), v2(n, 0);
        
        int res = 0;
        int min_val = a[0];
        for (int i = 0; i < n; i++) {
            res = max(res, a[i] - min_val);
            v1[i] = res;
            min_val = min(min_val, a[i]);
        }
        
        res = 0;
        int max_val = a[n-1];
        for (int i = n-1; i >= 0; i--) {
            res = max(res, max_val - a[i]);
            v2[i] = res;
            max_val = max(max_val, a[i]);
        }
        
        res = 0;
        for (int i = 0; i < n; i++)
            res = max(res, v1[i]+v2[i]);
        
        return res;
    }
};

File: python/123. Best Time to Buy and Sell Stock III.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        if n < 2: return 0
        dp1 = [0]*n # 表示i天之前的股票交易能取得的最大利润
        dp2 = [0]*n # 表示i天之后的股票交易能取得的最大利润

        min_val = prices[0]
        for i in range(1, n):
            dp1[i] = max(dp1[i-1], prices[i]-min_val)
            min_val = min(min_val, prices[i])

        max_val = prices[-1]
        for i in range(n-2, -1, -1):
            dp2[i] = max(dp2[i+1], max_val - prices[i])
            max_val = max(max_val, prices[i])
        
        return max([dp1[i]+dp2[i] for i in range(n)])
        